### PR TITLE
Array: Fix NEP18 dispatching in finalize

### PR DESCRIPTION
**Pull request by gjoseph92**
#4669 was meant to do this, but it wasn't working anymore. Turns out `arr.__array_function__ is not type(arr).__array_function__`, even when `arr` is a plain ndarray. So I believe we've actually being going down the `_concatenate2` path for all array types, even plain ndarrays where we could do the more efficient fill-in-the-blocks method of `concatenate3`.

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -4767,7 +4767,9 @@ def dtype(x):
 
     result = np.empty(shape=shape, dtype=dtype(deepfirst(arrays)))
 
-    for (idx, arr) in zip(slices_from_chunks(chunks), core.flatten(arrays)):
+    for (idx, arr) in zip(
+        slices_from_chunks(chunks), core.flatten(arrays, container=(list, tuple))
```
Comment (gjoseph92) [inline dask/array/core.py:4771]: Adding `container=(list, tuple)` to `flatten` here is a driveby. Seemed weird to me that we gave it earlier in the function but not here. Not sure if there's any reason for the inconsistency, so I made it consistent.

**Comment by gjoseph92** [comment]
FYI @jsignell I found this in the process of pulling the ndarray-subclass commits out https://github.com/dask/dask/pull/7491#issuecomment-812167610

**Review by jsignell — APPROVED** [review]
This looks good to me. Thanks!
